Add context for inlined functions and fix a number of other bugs

o- opened the pull request:
Increases max compilation unit size, which uncovered all those bugs addressed in this PR.

See the individual commits.

Replaces #336

o- committed:
Increase max compilation unit size and fix a number of bugs in the compiler

Due to large methods being excluded from compilation a number of
bugs managed to sneak in. They got revealed when I tried to
increase the limits. This commit fixes:

* the RIR push_ bytecode changed the visibility flag. However pir
LdConst of course is effectless. This can lead to missed invisible
settings. This is solved by making the visible explicit in rir already.
I fixed this, because it spams the test logs with, with NULL's from
dangling else branches in `cleanEx`, which lost their invisible.
* The above fix kills some optimizations. Thus i improved the visibility
analysis a bit. But also, We don't really care about
visible that much. We just want to get it right most of the times. Thus
this commit ignores visibility in some strategic cases. For example
promises which are pure except for visibility change, we still inline
(the subsequent call will change visibility anyway). Checkpoints also
ignore visibility. Worst case we deopt with the wrong visibility
setting.
* The scope analysis would sometimes generate pretty terrible phis, that
caused the lowering in the allocator to just diverge. Excluded those
kind of phis in the verifier and fixed the scope analysis to not produce
them.
* cleanup would happily merge the block with a phi input and the block
with the actual phi. this leads to input+phi ending up in the same
block.
* banned a couple more reflective functions from being inlined

o- committed:
fix parameter on travis

o- committed:
add contexts for inlined functions

If we inline functions, and those functions are not leaf functions (ie.
they have a call instruction in them), it is not sound to not allocate a
context (ie. a RCNTXT) for them.

If for example a callee of the inlinee does reflection on the inlinee,
it will see the context of the caller instead. Something like:

    Caller    -->     Inlinee   -->   CalleeOfInlinee
    [    just one context   ]   -->   [  sys.parent ]

In this case sys.parent returns the context of Caller instead of
Inlinee.

To solve this issue an explicit push_/pop_context instruction is added
to pir and rir. On the rir side we add an optimization to remove
contexts again, if after optimizations the inlinee does not have
outgoing calls anymore.

On the rir side we use an inlinedContextTrampoline (similar to the loop
context trampoline) that recursively calls the interpreter function,
while keeping the current interpreter state (like local variables, pc,
stack and so on). The corresponding pop_context terminates that
recursive invocation and continues execution in the original call frame.

charig commented on rir/src/compiler/analysis/query.cpp:
```diff
@@ -17,7 +17,9 @@ bool Query::noEnv(Code* c) {
 bool Query::pure(Code* c) {
     return Visitor::check(c->entry, [](Instruction* i) {
-        return !i->hasEffect() && !i->changesEnv();
+        // Yes visibility is a global effect. We try to preserve it. But geting
```
Great comment :). Still, I'd move it to the definition of `hasEffectIgnoreVisibility`. And put sth more like a sth like `TODO: devise a good story for supporting visibility without incuring in considerable overheads`

charig commented on rir/src/compiler/analysis/verifier.cpp:
```diff
@@ -176,32 +176,125 @@ class TheVerifier {
+                    if (input == phi->bb()) {
+                        std::cerr << "Error at instruction '";
+                        i->print(std::cerr);
+                        std::cerr << "': input '";
```
I do not get this new block.
1. Couldn't this happen in a loop block?
2. The error message: `input has the phi block as input block`. What is the input block of an argument? The block where it is defined? I'd say input lives in the same block as the phi or sth like that.

o- replied:
1. only if we allowed critical edges

charig commented on rir/src/compiler/analysis/visibility.cpp:
```diff
@@ -0,0 +1,65 @@
+            flag = getFlag(CallSafeBuiltin::Cast(i)->builtinId);
+        }
+
+        if (flag < 2) {
```
flag is only for visibility? If yes change to getVisibility. Otherwise, maybe for code readability instead of the `2` we can use a static mask defined somewhere for gathering the proper bits

o- replied:
this is R terminology, would rather not invent new names for it. see names.c

charig commented on rir/src/compiler/opt/inline.cpp:
```diff
@@ -97,6 +97,9 @@ class TheInliner {
                             return false;
                         }
                     }
+                    if (CallInstruction::CastCall(i)) {
```
So, we are finally not inlining any function that contains a call?

o- replied:
we are. read the next line. it's just that we need to add push/pop context instructions

charig commented on rir/src/compiler/translations/pir_2_rir/pir_2_rir.cpp:
```diff
@@ -879,24 +881,11 @@ size_t Pir2Rir::compileCode(Context& ctx, Code* code) {
                 break;
             }
 
-            case Tag::CastType: {
```
Good catch moving this to the verifier!

charig commented on rir/src/interpreter/interp.cpp:
```diff
@@ -11,6 +11,7 @@
 
 #include <assert.h>
 #include <deque>
+#include <set>
```
why?

o- replied:
`grep std::set rir/src/interpreter/interp.cpp` ?

charig commented on rir/src/interpreter/interp.cpp:
```diff
@@ -346,6 +346,46 @@ static void loopTrampoline(Code* c, InterpreterInstance* ctx, SEXP* env,
     Rf_endcontext(&cntxt);
 }
 
+const static SEXP inlineContextTrampolineDeoptMarker = (SEXP)0x7008;
```
inside?

charig commented on rir/src/compiler/analysis/unnecessary_contexts.h:
```diff
@@ -0,0 +1,56 @@
+
+    bool canRemove(PopContext* i) {
+        auto res = StaticAnalysis::at<PositioningStyle::BeforeInstruction>(i);
+        return res.get() && res.get() == i->push() && !res.needed;
```
the `res.get() seems a little confusing to me. Maybe an assert()?

o- replied:
doing a null test by relying on `nullptr <==> false` is very idiomatic. assert? we are not aborting if it's null, just return false.

charig commented on rir/src/compiler/opt/context.cpp:
```diff
@@ -0,0 +1,55 @@
+        while (ip != bb->end()) {
+            auto next = ip + 1;
+            auto instr = *ip;
+
```
e -> context?

Maybe an assert for even or 0 after the visitor?

charig commented on rir/src/compiler/pir/instruction.h:
```diff
@@ -690,11 +690,14 @@ class Branch
-        : FixedLenInstruction(PirType::voyd(), {{PirType::val()}}, {{ret}}) {}
+    explicit Return(Value* ret, Value* env)
+        : FixedLenInstructionWithEnvSlot(PirType::voyd(), {{PirType::val()}},
+                                         {{ret}}, env) {}
```
I do not remember anything in this commit related to the addition of the env here...

charig commented on rir/src/compiler/translations/pir_2_rir/pir_2_rir.cpp:
```diff
@@ -1018,6 +1020,22 @@ size_t Pir2Rir::compileCode(Context& ctx, Code* code) {
+
+            case Tag::PopContext: {
+                auto push = PopContext::Cast(instr)->push();
+                if (!pushContexts.count(push))
```
Is it possible that we still have not processed the push?

o- replied:
probably not. but didn't want to unnecessarily enforce a particular order to dump the basic blocks...

charig commented on rir/src/interpreter/interp.cpp:
```diff
@@ -346,6 +346,46 @@ static void loopTrampoline(Code* c, InterpreterInstance* ctx, SEXP* env,
+                       symbol::delayedArglist, op);
+
+    auto trampoline = [&]() {
+        if ((SETJMP(cntxt.cjmpbuf))) {
```
I have no idea but happens inside this if but it is also in the `rirCallTrampoline_` so ok...

o- replied:
stjmp-longjmp is used for non-local returns. You can jump from a nested function to this place. And you can also restart a function instead of just returning (used for continuations)

charig commented on rir/src/compiler/opt/inline.cpp:
```diff
@@ -257,7 +259,29 @@ class TheInliner {
                     });
 
-                    Value* inlineeRes = BBTransform::forInline(copy, split);
+                    auto inlineeRet = BBTransform::forInline(copy, split);
```
inlinee
inilineReturn
inlineReturnBB?

charig commented on rir/src/interpreter/interp.cpp:
```diff
@@ -3052,11 +3124,15 @@ SEXP evalRirCode(Code* c, InterpreterInstance* ctx, SEXP* env,
             SEXP res = ostack_pop(ctx);
-            SEXP e = ostack_pop(ctx);
+            // Inlined functions may or may not have contexts
+            if (e == R_GlobalContext->cloenv)
```
I do not clearly see why this condition couldn't hold while we did not create an context?

o- replied:
superseeded by changes

charig commented on rir/src/interpreter/instance.h:
```diff
@@ -172,14 +172,19 @@ class Locals final {
   private:
     R_bcstack_t* base;
     unsigned localsCount;
+    bool existingLocals;
```
Do not fully get this bool? Does it have any relation with inlining? When are the locals already reserved on the stack?

o- replied:
for inlinedContexts we recursively call the interpreter function, but we just continue executing as before. in particular we continue using the same locals and the stack. the additional interpreter execution is just to sneak a context trampoline in between, to bind the setjmps.

charig commented on rir/src/ir/BC.cpp:
```diff
@@ -100,6 +100,7 @@ BC_NOARGS(V, _)
     case Opcode::br_:
     case Opcode::brtrue_:
     case Opcode::beginloop_:
+    case Opcode::push_context_:
```
Why pop_contenxt_ is not considered in these switches?

o- replied:
it in the simple instruction macro that creates the entries...

charig reviewed (approved):
lgtm modulo ci

o- committed:
address review

o- committed:
fix inline context restart and disable flaky R test on DEOPT_CHAOS

o- committed:
cached recommended packages and don't fail if we can't sync them

o- committed:
fix benchmark tests

o- committed:
fix stub env insertion pass

o- committed:
fixing stack size on longjmp by patching the context

o- committed:
IsEnvStub needs to be tagged as instruction with environment

the delay env pass (and others) rely on the fact that all instructions
that refer to an environment have this flag set.

o- committed:
now gitlab times out :) parallelize more

o- committed:
disable crantools tests since we are currently blocked
